{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGCCGGAAAATGGGGCGGATAAGGAGTAGCTGAGTACTCCGCTGTCTTGGGGAGGATACCCATACGACGTGCCCGACTATGCAGTGTTAAATGAATAGGACGGGGGGCGGAAA\n",
      "\n",
      "M\n",
      "MG\n",
      "MGR\n",
      "MGRI\n",
      "MGRIR\n",
      "MGRIRS\n",
      "MGRIRSS\n",
      "MGRIRSS*\n",
      "M\n",
      "MQ\n",
      "MQC\n",
      "MQC*\n",
      "M\n",
      "MN\n",
      "MNR\n",
      "MNRT\n",
      "MNRTG\n",
      "MNRTGG\n",
      "MNRTGGG\n",
      "None\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "argument of type 'NoneType' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 201\u001b[0m\n\u001b[1;32m    198\u001b[0m out_loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/work/users/d/i/dieckhau/mRNA-display-data/tmp.csv\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    199\u001b[0m in_loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/work/users/d/i/dieckhau/mRNA-display-data/DRR337076.fastq\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m--> 201\u001b[0m \u001b[43mSelectionRoundSortedSequenceListGenerator\u001b[49m\u001b[43m(\u001b[49m\u001b[43mout_loc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43min_loc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    203\u001b[0m \u001b[38;5;66;03m# SelectionRoundSortedSequenceListGenerator('ES-Ace_R1_001.csv', 'ES-Ace_R1_001.fastq')\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[1], line 171\u001b[0m, in \u001b[0;36mSelectionRoundSortedSequenceListGenerator\u001b[0;34m(SortedPeptideSequecesFileName, fastqFileLocation)\u001b[0m\n\u001b[1;32m    168\u001b[0m SelectionRoundNumber \u001b[38;5;241m=\u001b[39m fastqFileLocation[fastqFileLocation\u001b[38;5;241m.\u001b[39mfind(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m    170\u001b[0m SortedSequenceFile \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(SortedPeptideSequecesFileName, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 171\u001b[0m SortedPeptideSequences \u001b[38;5;241m=\u001b[39m \u001b[43mSortedPeptideSequencesList\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfastqFileLocation\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    173\u001b[0m TotalPeptideNumber \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m    174\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m Data \u001b[38;5;129;01min\u001b[39;00m SortedPeptideSequences:\n",
      "Cell \u001b[0;32mIn[1], line 142\u001b[0m, in \u001b[0;36mSortedPeptideSequencesList\u001b[0;34m(fastqFileLocation)\u001b[0m\n\u001b[1;32m    140\u001b[0m Line \u001b[38;5;241m=\u001b[39m OpenReadingFrame(Line, minLen, maxLen)\n\u001b[1;32m    141\u001b[0m \u001b[38;5;28mprint\u001b[39m(Line)\n\u001b[0;32m--> 142\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\u001b[43mbeginning\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mLine\u001b[49m) \u001b[38;5;129;01mand\u001b[39;00m (end \u001b[38;5;129;01min\u001b[39;00m Line):\n\u001b[1;32m    143\u001b[0m     \u001b[38;5;28mprint\u001b[39m(Line)\n\u001b[1;32m    144\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'NoneType' is not iterable"
     ]
    }
   ],
   "source": [
    "#-------------------------------------------------------------------------------\n",
    "# utilise ONLY ONE stop codons, returns ONLY ONE ORF\n",
    "def OpenReadingFrame(DNASequence,minLen,maxLen):\n",
    "    '''\n",
    "    Finds first start codon, starts translating to peptide.\n",
    "    Stops and rejects sequence if upper length limit hit.\n",
    "    Stops translation at first in-frame stop codon.\n",
    "    Rejects peptide if below lower length limit.\n",
    "    If acceptable length peptide found, returns as tuple: (CDS, peptide)\n",
    "    Otherwise, truncates sequence to remove initial start codon, looks for the next one, then repeats translation attempt\n",
    "    '''\n",
    "    #setting up codon table\n",
    "    translation = {\n",
    "    \"TTT\": \"F\", \"TCT\": \"S\", \"TAT\": \"Y\", \"TGT\": \"C\",\n",
    "    \"TTC\": \"F\", \"TCC\": \"S\", \"TAC\": \"Y\", \"TGC\": \"C\",\n",
    "    \"TTA\": \"L\", \"TCA\": \"S\", \"TAA\": \"*\", \"TGA\": \"*\",\n",
    "    \"TTG\": \"L\", \"TCG\": \"S\", \"TAG\": \"*\", \"TGG\": \"W\",\n",
    "\n",
    "    \"CTT\": \"L\", \"CCT\": \"P\", \"CAT\": \"H\", \"CGT\": \"R\",\n",
    "    \"CTC\": \"L\", \"CCC\": \"P\", \"CAC\": \"H\", \"CGC\": \"R\",\n",
    "    \"CTA\": \"L\", \"CCA\": \"P\", \"CAA\": \"Q\", \"CGA\": \"R\",\n",
    "    \"CTG\": \"L\", \"CCG\": \"P\", \"CAG\": \"Q\", \"CGG\": \"R\",\n",
    "\n",
    "    \"ATT\": \"I\", \"ACT\": \"T\", \"AAT\": \"N\", \"AGT\": \"S\",\n",
    "    \"ATC\": \"I\", \"ACC\": \"T\", \"AAC\": \"N\", \"AGC\": \"S\",\n",
    "    \"ATA\": \"I\", \"ACA\": \"T\", \"AAA\": \"K\", \"AGA\": \"R\",\n",
    "    \"ATG\": \"M\", \"ACG\": \"T\", \"AAG\": \"K\", \"AGG\": \"R\",\n",
    "\n",
    "    \"GTT\": \"V\", \"GCT\": \"A\", \"GAT\": \"D\", \"GGT\": \"G\",\n",
    "    \"GTC\": \"V\", \"GCC\": \"A\", \"GAC\": \"D\", \"GGC\": \"G\",\n",
    "    \"GTA\": \"V\", \"GCA\": \"A\", \"GAA\": \"E\", \"GGA\": \"G\",\n",
    "    \"GTG\": \"V\", \"GCG\": \"A\", \"GAG\": \"E\", \"GGG\": \"G\"}\n",
    "    \n",
    "    \n",
    "    #setting start codon to look for\n",
    "    StartCodon = 'ATG'\n",
    "    \n",
    "    #making copy of input (need to trim to start codon without affecting final output indexing)\n",
    "    SubString = DNASequence\n",
    "    print(SubString)\n",
    "    #looping over every available start codon, from the start, until acceptable peptide found\n",
    "    while StartCodon in SubString:\n",
    "#        print 'found start codon'\n",
    "        #find first start codon\n",
    "        StartIndex = SubString.find(StartCodon)\n",
    "        #trim sequence to start at start codon\n",
    "        SubString = SubString[SubString.find(StartCodon):]\n",
    "        #set up empty peptide sequence\n",
    "        PeptideSequence = ''\n",
    "        #making copy of input (need to trim during translation without affecting position of start codon search)\n",
    "        translationString = SubString\n",
    "        #scanning down every three letter set\n",
    "        while len(translationString) > 3:\n",
    "#            print 'current string: ' +str(translationString[0:3])+' '+str(translationString[3:])\n",
    "#            print 'translated to: '+ str(translation[translationString[0:3]])\n",
    "            #adding codon translation to peptide sequence\n",
    "            if 'N' in translationString[0:3]:\n",
    "                PeptideSequence += 'X'\n",
    "            else:\n",
    "                PeptideSequence += translation[translationString[0:3]]\n",
    "            #trimming sequence by last translated codon\n",
    "            translationString = translationString[3:]\n",
    "#            print 'current peptide: '+str(PeptideSequence)\n",
    "            #if upper length limit hit, stop translating\n",
    "            print(PeptideSequence)\n",
    "            if (len(PeptideSequence)>maxLen):\n",
    "#                print '--length limit hit--'\n",
    "                return SubString[0:(len(PeptideSequence)*3)],PeptideSequence\n",
    "                break\n",
    "            #if in frame stop codon found, check if length is ok. Either accept and stop or reject and continue\n",
    "            if '+' in PeptideSequence:\n",
    "                if len(PeptideSequence)>minLen:\n",
    "                    return SubString[0:(len(PeptideSequence)*3)],PeptideSequence\n",
    "                    break\n",
    "                else:\n",
    "                    break\n",
    "            elif '*' in PeptideSequence:\n",
    "                if len(PeptideSequence)>minLen:\n",
    "                    return SubString[0:(len(PeptideSequence)*3)],PeptideSequence\n",
    "                    break\n",
    "                else:\n",
    "                    break\n",
    "        #remove first 3 bases(start codon) and continue search\n",
    "        SubString = SubString[3:]\n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "# read data file (.fastq extension)\n",
    "#RawDataFile = open(\"/Users/NikitaLoik/Documents/R6.fastq\", 'r')\n",
    "#Lines = RawDataFile.readlines()\n",
    "#RawDataFile.close\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "# creat a .CSV file with sequence list, DNA sequences and peptide sequences\n",
    "#DNASequenceFile = open('TestR6.csv', 'w')\n",
    "#SelectionRound = 6\n",
    "#SequenceCounter = 0\n",
    "#for Line in Lines:\n",
    "#    ORF = OpenReadingFrame(Line)\n",
    "#    if ORF != None:\n",
    "#        PeptideSequence = Translation(ORF)\n",
    "#        SequenceCounter += 1\n",
    "#        DNASequenceFile.write('selection round # ' + str(SelectionRound) + ',' +\n",
    "#                            'ORF ' + str(SequenceCounter) + ',' +\n",
    "#                            ORF + ',' +\n",
    "#                            PeptideSequence + '\\n')\n",
    "#DNASequenceFile.close \n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "# return a list of lists with peptide-sequences and their frequencies, sorted by frequency in descending order\n",
    "def SortedPeptideSequencesList(fastqFileLocation):\n",
    "    RawDataFile = open(fastqFileLocation, 'r')\n",
    "    Lines = RawDataFile.readlines()\n",
    "    RawDataFile.close\n",
    "    PeptideSequences = {}\n",
    "    PeptideSequencesList = []\n",
    "    \n",
    "    SelectionRoundNumber = fastqFileLocation[fastqFileLocation.find('.')-1]\n",
    "    \n",
    "    # populate the dictionary, so that Peptides are the keys and \n",
    "\n",
    "    #minLen = input(\"Minimum sequence length desired: \")\n",
    "    #maxLen = input(\"Maximum sequence length desired: \")\n",
    "    # minLen = 38\n",
    "    # maxLen = 42\n",
    "\n",
    "    # 5S5 (LazBF) primers (d.t.a.F95)\n",
    "    beginning = 'TTGCCGGAAAACGGGGCG' # codons for LPENGA\n",
    "    end = 'GGAGGATACCCATACGACGTGCCCGACTATGCAGTGTTAAATGAA' # codons for GGYPYDVPDYAVLNE\n",
    "    minLen = 114\n",
    "    maxLen = 116\n",
    "    \n",
    "    for Line in Lines:\n",
    "        if Line.startswith('@'): # skip non-seq lines\n",
    "            continue\n",
    "        # filter by expected beginning/end sequences\n",
    "        Line = OpenReadingFrame(Line, minLen, maxLen)\n",
    "        print(Line)\n",
    "        if (beginning in Line) and (end in Line):\n",
    "            print(Line)\n",
    "            break\n",
    "            # Line = OpenReadingFrame(Line,minLen,maxLen)\n",
    "            if Line != None:\n",
    "                ORF = Line[0]\n",
    "                #print str(ORF)\n",
    "                Peptide = Line[1]\n",
    "                #print str(Peptide)\n",
    "                if Peptide not in PeptideSequences:\n",
    "                    PeptideSequences[str(Peptide)] = 1\n",
    "                else:\n",
    "                    PeptideSequences[str(Peptide)] = PeptideSequences[str(Peptide)] + 1\n",
    "\n",
    "    # convert the dictionary into the list of lists\n",
    "    for key, value in PeptideSequences.items():\n",
    "        PeptideSequencesList.append([str(SelectionRoundNumber), key, value])\n",
    "    # sort the PeptideSequenceList by peptide sequence occurence in descendent order\n",
    "    SortedPeptideSequences = sorted(PeptideSequencesList, key = lambda x: x[2], reverse = True)\n",
    "    return SortedPeptideSequences\n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "#-------------------------------------------------------------------------------\n",
    "# creat a .CSV file with peptide-sequences list and their frequency, sorted by frequency in descending order\n",
    "def SelectionRoundSortedSequenceListGenerator(SortedPeptideSequecesFileName, fastqFileLocation):\n",
    "    \n",
    "    SelectionRoundNumber = fastqFileLocation[fastqFileLocation.find('.')-1]\n",
    "    \n",
    "    SortedSequenceFile = open(SortedPeptideSequecesFileName, 'w')\n",
    "    SortedPeptideSequences = SortedPeptideSequencesList(fastqFileLocation)\n",
    "    \n",
    "    TotalPeptideNumber = 0\n",
    "    for Data in SortedPeptideSequences:\n",
    "        TotalPeptideNumber = TotalPeptideNumber + Data[2]\n",
    "        \n",
    "    SortedSequenceFile.write('selection round # ' + str(SelectionRoundNumber) + '\\n' +\n",
    "                            'total sequence # ' + str(TotalPeptideNumber) + '\\n')\n",
    "                                            \n",
    "    UniqueSequenceNumber = 0\n",
    "    for Data in SortedPeptideSequences:\n",
    "        UniqueSequenceNumber += 1\n",
    "        PeptideSequence = Data[1]    \n",
    "        SequenceFraction = float(Data[2])/float(TotalPeptideNumber)\n",
    "        \n",
    "        SortedSequenceFile.write('ORF ' + str(UniqueSequenceNumber) + ',' +\n",
    "                                str(SelectionRoundNumber) + ',' +\n",
    "                                PeptideSequence + ',' +\n",
    "                                str(Data[2]) + ',' +\n",
    "                                '{:.3%}'.format(SequenceFraction) + '\\n')\n",
    "    SortedSequenceFile.close    \n",
    "#-------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#_____________________________RUNNING THE FUNCTION_____________________________#\n",
    "#___SortedPeptideSequecesFileName, fastqFileLocation___\n",
    "\n",
    "out_loc = '/work/users/d/i/dieckhau/mRNA-display-data/tmp.csv'\n",
    "in_loc = '/work/users/d/i/dieckhau/mRNA-display-data/DRR337076.fastq'\n",
    "\n",
    "SelectionRoundSortedSequenceListGenerator(out_loc, in_loc)\n",
    "\n",
    "# SelectionRoundSortedSequenceListGenerator('ES-Ace_R1_001.csv', 'ES-Ace_R1_001.fastq')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'B7H3b_ND_-A2R6.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[12], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msys\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mB7H3b_ND_-A2R6.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[0;32m      4\u001b[0m     file \u001b[38;5;241m=\u001b[39m f\u001b[38;5;241m.\u001b[39mreadlines()\n\u001b[0;32m      6\u001b[0m Lst1 \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:282\u001b[0m, in \u001b[0;36m_modified_open\u001b[1;34m(file, *args, **kwargs)\u001b[0m\n\u001b[0;32m    275\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[0;32m    276\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m    277\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    278\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    279\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    280\u001b[0m     )\n\u001b[1;32m--> 282\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m io_open(file, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'B7H3b_ND_-A2R6.csv'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "with open ('B7H3b_ND_-A2R6.csv') as f:\n",
    "    file = f.readlines()\n",
    "    \n",
    "Lst1 = []\n",
    "\n",
    "for row in file:\n",
    "    Lst1.append(row.split(','))\n",
    "\n",
    "Lst = []  \n",
    "counter = 0\n",
    "\n",
    "for line in Lst1:\n",
    "    counter += 1\n",
    "    x = counter\n",
    "    y = line[2]\n",
    "    z = line[3]\n",
    "    n = line[4]\n",
    "    Lst. append(\">\"+str(n))\n",
    "    Lst. append(y)\n",
    "    \n",
    "orig_stdout = sys.stdout\n",
    "Output = open(\"B7H3b_ND_-A2R6.fasta\", 'w')\n",
    "sys.stdout = Output\n",
    "\n",
    "for c in Lst:\n",
    "    print (c)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TOTAL RAW READS: 3517658\n",
      "READS OF CORRECT LENGTH: 3351776\n",
      "(array([ 36,  39,  42,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,\n",
      "        55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,\n",
      "        68,  69,  70,  71,  72,  75,  79,  81,  82,  83,  84,  85,  86,\n",
      "        87,  88,  90,  91,  92,  93,  94,  95,  96,  97,  98,  99, 100,\n",
      "       101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113,\n",
      "       114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126,\n",
      "       127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 139, 140, 141,\n",
      "       142, 145, 146, 147, 148, 149, 150, 151, 152]), array([      2,       2,       8,       8,       4,       5,      17,\n",
      "            16,      11,      40,      22,      55,     123,      16,\n",
      "            14,      24,      36,      17,      23,      10,     200,\n",
      "           134,      32,      39,      30,       7,       2,      13,\n",
      "             1,       2,       2,       1,       1,       6,       2,\n",
      "             5,       2,       4,       1,       3,      13,       3,\n",
      "            27,       7,       2,      39,       8,       9,      73,\n",
      "            16,      18,     300,      30,      21,     516,      48,\n",
      "            44,     779,     117,      84,    1511,     264,     241,\n",
      "          4611,    4308,   78499, 3351776,    7097,      98,     335,\n",
      "            26,      15,      67,      13,       8,      25,       5,\n",
      "             6,      24,       7,      13,      14,      12,       3,\n",
      "            35,       1,       1,      10,       6,       2,       3,\n",
      "             2,       3,       2,      76,      25,      11,      92,\n",
      "          3516,   61801]))\n",
      "115 best length\n",
      "(3351776, 4)\n"
     ]
    }
   ],
   "source": [
    "# Henry's attempt\n",
    "\n",
    "in_file = '/work/users/d/i/dieckhau/mRNA-display-data/DRR337076.fastq'\n",
    "out_file = '/work/users/d/i/dieckhau/mRNA-display-data/tmp.csv'\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def read_fastq(file, params):\n",
    "    with open(file, 'r') as fopen:\n",
    "        lines = fopen.readlines()\n",
    "\n",
    "    lines = np.array(lines) # L x 4\n",
    "    lines = np.reshape(lines, (lines.shape[0] // 4, 4)) # [L, 4]\n",
    "    # 0 is header\n",
    "    # 1 is seq\n",
    "    # 2 is duplicate header\n",
    "    # 3 is quality score\n",
    "    print('TOTAL RAW READS:', lines.shape[0])\n",
    "    \n",
    "    # filter by length\n",
    "    lengths = np.array([len(l) for l in lines[:, 1]])\n",
    "    \n",
    "    lines = lines[lengths == PARAMS['RAW_LENGTH'], :]\n",
    "    print('READS OF CORRECT LENGTH:', lines.shape[0])\n",
    "    \n",
    "    u, c = np.unique(lengths, return_counts=True)\n",
    "    print(np.unique(lengths, return_counts=True))\n",
    "    print(u[np.argmax(c)], 'best length')\n",
    "    print(lines.shape)\n",
    "    \n",
    "\n",
    "    return\n",
    "\n",
    "PARAMS = {\n",
    "    'RAW_LENGTH': 115\n",
    "}\n",
    "\n",
    "read_fastq(in_file, PARAMS)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
